### PR TITLE
[model] feat: support async ulysses for qwen3vl dense

### PR DESCRIPTION
**Pull request by yyytiancai**
### What does this PR do?

support async ulysses for qwen3vl dense

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by yyytiancai** [comment]
您好，已收到。

**Comment by yyytiancai** [comment]
label ascend

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py [hunk 1]
```diff
@@ -455,36 +461,90 @@ def forward(
         input_shape = hidden_states.shape[:-1]
         hidden_shape = (*input_shape, -1, self.head_dim)
 
-        query_states = self.q_norm(self.q_proj(hidden_states).view(hidden_shape)).transpose(1, 2)
-        key_states = self.k_norm(self.k_proj(hidden_states).view(hidden_shape)).transpose(1, 2)
-        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        if self.async_ulysses:
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: please follow the standard practice to mark any changes from HF with comment lile `# Modification: xxxxx` so that it's easier for future readers to scan over what are additional changes that veomni adds on top of HF code.

File: veomni/distributed/sequence_parallel/async_ulysses.py [hunk 2]
```diff
@@ -65,9 +73,29 @@ def forward(
         group: ProcessGroup,
     ):
         sp_group = get_ulysses_sequence_parallel_group() if group is None else group
+        ulysses_size = get_parallel_state().ulysses_size
```
Comment (piyifan123) [inline veomni/distributed/sequence_parallel/async_ulysses.py]: nit: please add docstrings for all public functions in this file.

also, it would be nice to add a doc page for async ulysses on 1) how it works and 2) how to modify a model code to support it.

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (piyifan123) [reply veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: please fix?

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py [hunk 3]
```diff
@@ -455,36 +461,90 @@ def forward(
...
+            unpadded_seq_len = hidden_states.size(1)
+            q, k, v = async_ulysses_qkv_projection(
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: not sure if a better idea is to fully encapsulate the async ulysses attn impl as a single function call?

Comment (piyifan123) [inline veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: do we have unit tests for those interfaces? we should have tests for those complex ops IMO as it involves complex computation/communication overlapping logic, having unit tests would make debugging, future optimization, and correctness verification much easier and forgiving in the future.

I'd suggest to add unit tests at first to fully cover the correctness of async_ulysses components and then start to add support in models.

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yyytiancai) [reply veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: self.async_ulysses was a temporary solution, it has been updated to read the async state through model training configuration parameters

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py (hunk 3 above)
Comment (yyytiancai) [reply veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: > not sure if a better idea is to fully encapsulate the async ulysses attn impl as a single function call?

done

Comment (yyytiancai) [reply veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py]: The related test cases: tests/parallel/ulysses/test_async_ulysses.py
Currently, this PR has been adapted to this test, and the test can pass

File: tasks/omni/train_qwen_vl.py [hunk 4]
```diff
@@ -117,6 +117,7 @@ def main():
         cp_size=args.train.context_parallel_size,
         ulysses_size=args.train.ulysses_parallel_size,
         dp_mode=args.train.data_parallel_mode,
+        async_ulysses=args.train.async_ulysses,
```
Comment (FoolPlayer) [inline tasks/omni/train_qwen_vl.py]: it's better enable async or non-async ulysses through args.model.xxx, (like args.model.ulysses_implementation = 'eager' or 'async'), please refer the use of `args.model.moe_implementation` https://github.com/ByteDance-Seed/VeOmni/blob/b8dcc17e332f042453e0ffaed5910439f2841e5f/veomni/models/auto.py#L90

Comment (Coach257) [reply tasks/omni/train_qwen_vl.py]: I think it's fine to keep it in `parallel_state`.

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py [hunk 5]
```diff
@@ -442,49 +447,131 @@ def __init__(self, config: Qwen3VLTextConfig, layer_idx: int):
...
         attn_output, attn_weights = attention_interface(
             self,
             query_states,
             key_states,
-            value_states,
+            v,
             attention_mask,
             dropout=0.0 if not self.training else self.attention_dropout,
             scaling=self.scaling,
+            skip_ulysses=True,
```
Comment (Coach257) [inline veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py:506]: del this. Move this to flash_attention_forward:
if sp_enabeld and not skip_ulysses and not async_enabled

File: tasks/omni/train_qwen_vl.py [hunk 6]
```diff
@@ -127,6 +127,7 @@ def main():
         force_use_huggingface=args.model.force_use_huggingface,
         moe_implementation=args.model.moe_implementation,
         attn_implementation=args.model.attn_implementation,
+        ulysses_implementation=args.model.ulysses_implementation,
```
Comment (Coach257) [inline tasks/omni/train_qwen_vl.py]: this arg seems useless. Use parallel_state.asysn_enabled is easier to control model forward

File: veomni/utils/arguments.py [hunk 7]
```diff
@@ -97,6 +97,10 @@ class ModelArguments:
         default=False,
         metadata={"help": "Force loading model from huggingface."},
     )
+    ulysses_implementation: Optional[Literal[None, "eager", "async"]] = field(
+        default="eager",
+        metadata={"help": "Ulysses parallelism execution mode."},
```
Comment (Coach257) [inline veomni/utils/arguments.py]: please use `async_enabled`

**Comment by Coach257** [comment]
Please rebase master

File: tasks/omni/train_qwen_vl.py (hunk 4 above)
Comment (yyytiancai) [reply tasks/omni/train_qwen_vl.py]: > I think it's fine to keep it in `parallel_state`.

done

File: tasks/omni/train_qwen_vl.py (hunk 6 above)
Comment (yyytiancai) [reply tasks/omni/train_qwen_vl.py]: done

File: veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py (hunk 5 above)
Comment (Coach257) [reply veomni/models/transformers/qwen3_vl/modeling_qwen3_vl.py:506]: keep `skip_ulysses` for special cases (vision attention not async but text attention sync)

**Comment by yyytiancai** [comment]
> Please rebase master

done

**Review by Coach257 — APPROVED** [review]
